### PR TITLE
Rd 12757 java tracer support kafka instrumentation

### PR DESCRIPTION
**Pull request by nadav3396**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -80,7 +80,7 @@
         <dependency>
             <groupId>com.amazonaws</groupId>
             <artifactId>aws-lambda-java-events</artifactId>
-            <version>2.2.6</version>
+            <version>3.11.5</version>
```
Comment (nadav3396) [inline pom.xml:83]: needed for KafkaEvent trigger by

File: src/main/java/io/lumigo/core/instrumentation/impl/ApacheKafkaProducerInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            private final ProducerRecord<K, V> record;
+            private final long startTime;
+
+            @Override
+            public void onCompletion(RecordMetadata recordMetadata, Exception exception) {
+                try {
+                    if (callback != null) {
+                        callback.onCompletion(recordMetadata, exception);
+                    }
+                    Logger.info("Handling kafka request {}", record.hashCode());
```
Comment (moshe-shaham-lumigo) [inline src/main/java/io/lumigo/core/instrumentation/impl/ApacheKafkaProducerInstrumentation.java:103]: will this be printed in the default log level?

File: src/main/java/io/lumigo/core/instrumentation/impl/ApacheKafkaProducerInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                                                        takesArgument(
+                                                                1,
+                                                                named(
+                                                                        "org.apache.kafka.clients.producer.Callback")))),
+                        ApacheKafkaProducerAdvice.class.getName());
+    }
+
+    public static class ApacheKafkaProducerAdvice {
+        public static final SpansContainer spansContainer = SpansContainer.getInstance();
+        public static final LRUCache<Integer, Boolean> handled = new LRUCache<>(1000);
```
Comment (moshe-shaham-lumigo) [inline src/main/java/io/lumigo/core/instrumentation/impl/ApacheKafkaProducerInstrumentation.java:56]: what's the purpose of this?

File: src/main/java/io/lumigo/core/instrumentation/impl/ApacheKafkaProducerInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            private final long startTime;
+
+            @Override
+            public void onCompletion(RecordMetadata recordMetadata, Exception exception) {
+                try {
+                    if (callback != null) {
+                        callback.onCompletion(recordMetadata, exception);
+                    }
+                    Logger.info("Handling kafka request {}", record.hashCode());
+                    spansContainer.addKafkaProduceSpan(
```
Comment (moshe-shaham-lumigo) [inline src/main/java/io/lumigo/core/instrumentation/impl/ApacheKafkaProducerInstrumentation.java:104]: i'm not sure if it's good to have the spancontainer call the factory. maybe it's best to first create the span in the instrumentation and then add to the span container

File: src/main/java/io/lumigo/models/KafkaSpanFactory.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                                keySerializer.serialize(
+                                        record.topic(), record.headers(), record.key()))
+                        .value(
+                                valueSerializer.serialize(
+                                        record.topic(), record.headers(), record.value()))
+                        .headers(extractHeaders(record.headers()))
+                        .build();
+
+        KafkaSpan.KafkaInfo info;
+        if (exception == null) {
```
Comment (moshe-shaham-lumigo) [inline src/main/java/io/lumigo/models/KafkaSpanFactory.java:47]: there is some code duplication in this if statement. it can be improved

File: src/main/java/io/lumigo/models/KafkaSpanFactory.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                            .topic(topic)
+                            .record(producerRecord)
+                            .response(
+                                    KafkaSpan.KafkaProducerErrorResponse.builder()
+                                            .errorMessage(exception.getMessage())
+                                            .build())
+                            .build();
+        }
+
+        Header messageIdHeader = record.headers().lastHeader(KafkaSpan.LUMIGO_MESSAGE_ID_KEY);
```
Comment (moshe-shaham-lumigo) [inline src/main/java/io/lumigo/models/KafkaSpanFactory.java:74]: so if i understand correctly, every kafka span holds a single message id?
what about batch requests?

File: src/main/java/io/lumigo/core/utils/AwsUtils.java [hunk 7]
```diff
@@ -155,6 +154,25 @@ public static TriggeredBy extractTriggeredByFromEvent(Object event) {
...
+                triggeredBy.setTriggeredBy("kafka");
+                triggeredBy.setArn(((KafkaEvent) event).getEventSourceArn());
+                List<String> messageIds = new ArrayList<>();
+                for (Map.Entry<String, List<KafkaEvent.KafkaEventRecord>> entry :
+                        ((KafkaEvent) event).getRecords().entrySet()) {
+                    for (KafkaEvent.KafkaEventRecord record : entry.getValue()) {
+                        for (Map<String, byte[]> headers : record.getHeaders()) {
+                            if (headers.containsKey(KafkaSpan.LUMIGO_MESSAGE_ID_KEY)) {
+                                messageIds.add(
+                                        new String(
```
[the remark this answers is not in the document]
Comment (nadav3396) [reply src/main/java/io/lumigo/core/utils/AwsUtils.java:167]: kafka headers are encoded to bytes and message ids are strings

File: src/main/java/io/lumigo/core/instrumentation/impl/ApacheKafkaProducerInstrumentation.java (hunk 2 above)
Comment (nadav3396) [reply src/main/java/io/lumigo/core/instrumentation/impl/ApacheKafkaProducerInstrumentation.java:103]: no, only with lumigo debug

File: src/main/java/io/lumigo/core/SpansContainer.java [hunk 8]
```diff
@@ -38,13 +47,15 @@ public class SpansContainer {
...
     private static final String HTTP_SPAN_TYPE = "http";
+    public static final String KAFKA_SPAN_TYPE = "kafka";
 
     private Span baseSpan;
     private Span startFunctionSpan;
     private Long rttDuration;
     private Span endFunctionSpan;
     private Reporter reporter;
     private List<HttpSpan> httpSpans = new LinkedList<>();
+    private List<KafkaSpan> kafkaSpans = new LinkedList<>();
```
[the remark this answers is not in the document]
Comment (moshe-shaham-lumigo) [reply src/main/java/io/lumigo/core/SpansContainer.java:58]: i think it makes sense.
is that what we do in other tracers?

File: src/main/java/io/lumigo/core/utils/AwsUtils.java [hunk 9]
```diff
@@ -282,10 +308,44 @@ private static String extractMessageIdFromDynamodbRecord(
             DynamodbEvent.DynamodbStreamRecord record) {
         if (record.getEventName() == null) return null;
         if (record.getEventName().equals("INSERT")) {
-            return dynamodbItemToHash(record.getDynamodb().getNewImage());
+            return calculateItemHash(record.getDynamodb().getNewImage());
```
Comment (moshe-shaham-lumigo) [inline src/main/java/io/lumigo/core/utils/AwsUtils.java:311]: what is the purpose of these changes?

File: src/main/java/io/lumigo/models/BaseSpan.java [hunk 10]
```diff
@@ -0,0 +1,3 @@
+package io.lumigo.models;
+
+public interface BaseSpan {}
```
Comment (moshe-shaham-lumigo) [inline src/main/java/io/lumigo/models/BaseSpan.java:3]: don't you want to move common fields to this base span?

started, ended, id, etc.

**Review by moshe-shaham-lumigo — APPROVED** [review]

File: src/main/java/io/lumigo/core/utils/AwsUtils.java (hunk 9 above)
Comment (nadav3396) [reply src/main/java/io/lumigo/core/utils/AwsUtils.java:316]: when i updated the lambda event library this got broken because in the new version they created the same object but inside this library (before they import it from dynamodb package)
so now the old logic isnt working becuase its different classes (with the same name, but methods with different names) so i needed to implement it again
